### PR TITLE
Subject: Support to include ID/PK in validation result for each row t…

### LINKED ISSUE #3195: Support to include ID/PK in validation result for each row that failed an expectations
**Is your feature request related to a problem? Please describe.**
When an expectation is run the output includes a “partial_unexpected_list” property of values that were unexpected. While this is useful, in most cases, it doesn't allow teams to identify, resolve, or divert data with poor quality.

It would be great if a sample list or a complete list (result_format SUMMARY, COMPLETE) of ID's for each row that failed an expectation was included in the validation result.

**Describe the solution you'd like**
One could include a column name as an argument in the expectation that they would like to be included in the "partial_unexpected_list" (or a new property).

**Describe alternatives you've considered**
One could try infer the PK for a table but this is not possible for all engine types E.g. Spark. 

**Additional context**
Enabling teams to not only bring light to data quality issues but identify all rows allows them to address poor data quality in real-time instead of requiring manual intervention.

### PR DESCRIPTION
**Pull request by abekfenn**
…hat failed an expectations

Ticket Number: #3195

Problem:

- No arguments available to configure column subset of unexpected_rows

Solution:

- Add args to subset unexpected_rows

**Updated 2022-11-10 by @Shinnnyshinshin**

Changes proposed in this pull request:

Introduction of `unexpected_index_columns` value that allows users to specify a primary key (PK) column for identifying rows that failed an Expectation (usually returned as part of the `unexpected_index_list`).
 
- Changes were made to the `map_metric_provider` to take in the parameter from `result_format` and output `unexpected_index_list` as key-value pairs of the primary key column. 

* **Note** This is only the `Pandas` implementation, with `Spark` and `sql` to follow. 

## Can you give me an Example? 
Given the following DataFrame: 

```python 
df = pd.DataFrame(
        {
            "pk_1": [0, 1, 2, 3, 4, 5],
            "pk_2": ["zero", "one", "two", "three", "four", "five"],
            "numbers_with_duplicates": [1, 5, 22, 3, 5, 10],
            "animal_names_no_duplicates": [
                "cat",
                "fish",
                "dog",
                "giraffe",
                "lion",
                "zebra",
            ],
        }
    )

```

We could run the `expect_column_values_to_be_in_set` Expectation on the `numbers_with_duplicates` column (which has `[1, 5, 22, 3, 5, 10]` as its values) with  `[1, 5, 22]` as the `value_set`.

```python
    expectationConfiguration = ExpectationConfiguration(
        expectation_type="expect_column_values_to_be_in_set",
        kwargs={
            "column": "numbers_with_duplicates",
            "value_set": [1, 5, 22],
        },
    )
```

After running the ExpectationConfiguration, we would expect the `unexpected_index_list` to be `[3, 5]` which correspond to the indices of `3` and `10`, the values  that are not in the `value_set`. 

This PR enables the following configuration, which sets `unexpected_index_columns` to be `pk_2`. 

```python

    expectationConfiguration = ExpectationConfiguration(
        expectation_type="expect_column_values_to_be_in_set",
        kwargs={
            "column": "numbers_with_duplicates",
            "value_set": [1, 5, 22],
            "result_format": {
                "unexpected_index_columns": ["pk_2"],
            },
        },
    )
```

After running this new ExpectationConfiguration, we expect the `unexpected_index_list` to be `[{"pk_2": "three"}, {"pk_2": "five"}]` which correspond to the values in the `pk_2` column of the indices of `22` and `10` the `numbers_with_duplicates` column. 

**Note**
* The PR also allows specifying multiple `unexpected_index_columns`, which are described in the tests. 
* The PR also allows for the output to follow patterns established by existing `result_format` levels (ie. `SUMMARY`, `COMPLETE`, `BASIC` and `BOOLEAN_ONLY`. 
- Duplicated by PR #6329 
- Closes #3195 

### Definition of Done
- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/expectations/expectation.py [hunk 1]
```diff
@@ -1861,16 +1865,16 @@ def get_validation_dependencies(
         if result_format_str in ["BASIC", "SUMMARY"]:
             return dependencies
 
-        if include_unexpected_rows:
+        if include_unexpected_index_columns:
             metric_kwargs = get_metric_kwargs(
-                f"{self.map_metric}.unexpected_rows",
+                f"{self.map_metric}.include_unexpected_index_columns",
```
Comment (austiezr) [inline great_expectations/expectations/expectation.py]:
```suggestion
                f"{self.map_metric}.unexpected_index_columns",
```

File: great_expectations/expectations/expectation.py [hunk 2]
```diff
@@ -1861,16 +1865,16 @@ def get_validation_dependencies(
...
                 configuration=configuration,
                 runtime_configuration=runtime_configuration,
             )
             metric_dependencies[
-                f"{self.map_metric}.unexpected_rows"
+                f"{self.map_metric}.include_unexpected_index_columns"
```
Comment (austiezr) [inline great_expectations/expectations/expectation.py]:
```suggestion
                f"{self.map_metric}.unexpected_index_columns",
```

File: great_expectations/expectations/expectation.py [hunk 3]
```diff
@@ -1861,16 +1865,16 @@ def get_validation_dependencies(
...
             ] = MetricConfiguration(
-                metric_name=f"{self.map_metric}.unexpected_rows",
+                metric_name=f"{self.map_metric}.include_unexpected_index_columns",
```
Comment (austiezr) [inline great_expectations/expectations/expectation.py]:
```suggestion
                f"{self.map_metric}.unexpected_index_columns",
```

File: tests/expectations/metrics/test_map_metric.py [hunk 4]
```diff
@@ -270,6 +273,53 @@ def test_pandas_unexpected_rows_complete_result_format(dataframe_for_unexpected_
...
+        "partial_unexpected_index_list": [3, 5],
+        "partial_unexpected_counts": [
+            {"value": 3, "count": 1},
+            {"value": 10, "count": 1},
+        ],
+        "missing_count": 0,
+        "missing_percent": 0.0,
+        "unexpected_percent_total": 33.33333333333333,
+        "unexpected_percent_nonmissing": 33.33333333333333,
+    }
```
Comment (austiezr) [inline tests/expectations/metrics/test_map_metric.py]: This assert will need a review once the above metric changes are in place.

**Review by austiezr — CHANGES_REQUESTED** [review]
Hey @abekfenn! Thanks for the work here; love to see it! Looks like you're very close to having this functional; just a couple of small fixes and I think you'll be able to move ahead with this.

**Comment by austiezr** [comment]
Looks like we'll need a linting pass as well!

**Comment by abekfenn** [comment]
Thanks @austiezr your feedback was very helpful. I'm definitely much closer now and was looking to add a test for an expectation that leverages the `SqlAlchemyExecutionEngine` but I don't see any existing tests or datasets that I could reference.
I assume they're out there, would you be able to point me to an example test or dataset that I could use to test an expectation with these args?

**Comment by austiezr** [comment]
Hey @abekfenn ! Have you had an opportunity to check out [these tests](https://github.com/great-expectations/great_expectations/blob/develop/tests/execution_engine/test_sqlalchemy_execution_engine.py)?

**Comment by abekfenn** [comment]
@austiezr I haven't seen those tests but I was looking for tests that test metric providers and the results that different configurations and providers produce. Such as those in `tests/expectations/metrics/test_map_metric.py`. I'll see if I can put something together from both test files, but surprised there are no tests that test the results produced by the metric providers for SQLAlchemy datasets.

**Comment by Shinnnyshinshin** [comment]
Thank you very very much @abekfenn for getting the ball rolling, and all the work you put in. I was able to start where you left off, and have a PR for the Pandas implementation that is currently under review by the team. I'll post more updates here :)

File: great_expectations/expectations/metrics/map_metric_provider.py [hunk 5]
```diff
@@ -1733,10 +1733,31 @@ def _pandas_map_condition_index(
...
+                if column_name not in metrics["table.columns"]:
+                    raise ge_exceptions.InvalidMetricAccessorDomainKwargsKeyError(
+                        message=f'Error: The unexpected_index_column: "{column_name}" in does not exist in Dataframe. '
+                        f"Please check your configuration and try again."
+                    )
+                primary_key_dict[column_name] = df.at[index, column_name]
+            unexpected_index_list.append(primary_key_dict)
+        if result_format["result_format"] == "COMPLETE":
+            return unexpected_index_list
+        return unexpected_index_list[: result_format["partial_unexpected_count"]]
```
Comment (Shinnnyshinshin) [inline great_expectations/expectations/metrics/map_metric_provider.py:1755]: From #6329 

Added the filtering for `partial_unexpected_count` for both `unexpected_index_column` and just `unexpected_index` (existing behavior)

File: tests/expectations/metrics/test_map_metric.py [hunk 6]
```diff
@@ -334,6 +352,403 @@ def test_pandas_unexpected_rows_complete_result_format(dataframe_for_unexpected_
...
+
+
+def test_pandas_multiple_unexpected_index_column_names_summary_result_format_limit_1(
+    pandas_dataframe_for_unexpected_rows_with_index: pd.DataFrame,
+):
```
Comment (Shinnnyshinshin) [inline tests/expectations/metrics/test_map_metric.py:605]: From #6329 

1st test for testing `partial_unexpected_count` limit

File: tests/expectations/metrics/test_map_metric.py [hunk 7]
```diff
@@ -334,6 +352,403 @@ def test_pandas_unexpected_rows_complete_result_format(dataframe_for_unexpected_
...
+        "missing_count": 0,
+        "missing_percent": 0.0,
+        "unexpected_percent_total": 33.33333333333333,
+        "unexpected_percent_nonmissing": 33.33333333333333,
+    }
+
+
+def test_pandas_multiple_unexpected_index_column_names_complete_result_format_limit_1(
+    pandas_dataframe_for_unexpected_rows_with_index: pd.DataFrame,
+):
```
Comment (Shinnnyshinshin) [inline tests/expectations/metrics/test_map_metric.py:511]: From https://github.com/great-expectations/great_expectations/pull/6329

2nd test for testing partial_unexpected_count limit

**Review by anthonyburdi — APPROVED** [review]
LGTM! Thank you for these changes 🙇

**Review by austiezr — APPROVED** [review]
❤️ 🚀 LGTM!

**Comment by abekfenn** [comment]
Thanks a lot for wrapping up the changes on these @Shinnnyshinshin! 

I only see changes related to pandas but my understanding was that this feature (stemming from https://github.com/great-expectations/great_expectations/issues/3195#) is really needed in the SQL engine in order to support identification of the Pk/Id of a failed row.
This functionality already exists for pandas in GE but is not supported for SQL.
Just want to make sure that merging doesn't mean that this issue will be closed out and that we will continue work on this issue for the SQL engine implementation?

**Comment by Shinnnyshinshin** [comment]
Hi @abekfenn thank you for the follow-up message 😄  Yes the next steps for us are to do the implementation for SQL and Spark (we'll be sure to do SQL first).  Wanting to thank you again for helping us get this far, and Ill be sure to update you once the SQL implementation has been implemented and merged.

**Comment by abekfenn** [comment]
Awesome! Thanks so much @Shinnnyshinshin

**Pull request merged**